proposals/0004-open-api.md

```diff
@@ -0,0 +1,322 @@
+        }
+        post("/users") {
+            userService.createUser(call.receive())
+            call.respond(HttpStatusCode.Created)
+        }
+    }
+}
+```
+
+As each route in the example is defined from the DSL, it is added to the application's internal model.  We can use this model to infer details pertaining to each route:
+1. The merged path
+2. The path parameters
+3. The HTTP method
+
+Because the handling of parameters and responses is contained to the route's lambda argument, we cannot infer details about them for the specification.  As a result, the default model will appear very basic without additional sources.
+
+To address this requirement, we indent to supplement the path information with a KDoc-like annotation API that can be read by our Gradle plugin.
+
+## Annotation API
+[annotation-api]: #annotation-api
+
+The annotation API provides a non-intrusive way to enhance the OpenAPI specification with details that cannot be inferred from code.
+
+Each endpoint will need to be annotated with a KDoc comment that follows this general format:
+
+```kotlin
+/**
+ * A summary of the endpoint
+ * 
+ * @<key> <value>
+ * ...
+ */
+get("/widgets") {
+    call.respond(widgetService.list())
+}
+```
+
+Here is a list of the parameters to be supported:
+
+| Tag            | Format                              | Description                                                  |
+|----------------|-------------------------------------|--------------------------------------------------------------|
+| `@tag`         | `@tag [TagName]`                    | Associates the endpoint with a tag for grouping              |
+| `@param`       | `@param name description`           | Describes a path or query parameter                          |
+| `@header`      | `@header name description`          | Describes a header parameter                                 |
+| `@cookie`      | `@cookie name description`          | Describes a cookie parameter                                 |
+| `@body`        | `@body [Type] description`          | Documents the request body type                              |
+| `@response`    | `@response code [Type] description` | Documents a response code with optional type and description |
+| `@deprecated`  | `@deprecated reason`                | Marks an endpoint as deprecated                              |
+| `@description` | `@description text`                 | Provides a detailed endpoint description                     |
+| `@security`    | `@security scheme`                  | Documents security requirements                              |
+
+Note that for `@body` and `@response` fields, a type reference is specified, which will be used to automatically supply the schema definition in the OpenAPI specification.
+
+By default, all parameters will be considered required unless the name is suffixed with a question mark (`?`).  For example, `@param name?` indicates that the parameter is optional.
```

Comment:
Should we include `default` as well as `optional`?
https://swagger.io/docs/specification/v3_0/describing-parameters/#default-parameter-values

Reply:
In addition to it. Should we also add information about parameter type and value restrictions (range, regex, etc.)?